bip-annex.mediawiki

```diff
@@ -0,0 +1,173 @@
+<pre>
+  BIP: XXX
+  Layer: Consensus (soft fork)
```

Comment:
I thought the taproot upgrade introduced the annex field to allow for future protocol expansions without requiring further soft forks? Doesn't the requirement of this bip for another soft fork to make the field useable defeat that purpose?

Reply:
The annex field gives space for future protocol expansions, but each expansion requires a soft fork in order to give it semantics.

That said, this particular format BIP in principle could alternatively be defined so that invalidly formatted annex fields do not invalidate the transaction, and instead just prevent the interpretation of the annex in this format.  There would be advantages and disadvantages to such an alternative definition.

Reply:
There could also be a policy-only relaxation(as we've done in the inquisition repo), but the same tradeoffs apply

Reply:
> I thought the taproot upgrade introduced the annex field to allow for future protocol expansions without requiring further soft forks? Doesn't the requirement of this bip for another soft fork to make the field useable defeat that purpose?

The idea is to have a type-length-value record where new semantics can be assigned to each record without the record developers having to think about the consensus syntax issues (e.g what if another record use multi-bytes length) though the semantic issues will be still something to reason on.

I still think even if you have a policy-only relaxation of the annex, we have to deploy a relaxation for each new record.

Reply:
Thinking about the most minimal change to make the annex useable on the short term. How about just defining in policy a way to express an 'unstructured' block of data for now? For example starting unstructured data with byte `0`, and then later define tlv or something else in a way that the first byte is never `0`?

Reply:
This is probably also more efficient for the unstructured data case than tlv using a single record that still requires a length to be encoded.

Reply:
> Thinking about the most minimal change to make the annex useable on the short term. How about just defining in policy a way to express an 'unstructured' block of data for now?

Looking on current approach we would burn at least one type record (`0x00`) ? This is more efficient for unstructured data as effectively you save the encoded length. As an alternative we can versions the annex and have one of them being reserved for policy-only experimentations.

Concerns with current approach if/when consensus semantic is assigned to the annex, all the policy-only annexes will become enforced in a different way, breaking use-cases. Protocol evolvability matters.

Reply:
Unstructured data has a lot of uses. With nothing in the annex having consensus meaning yet, these are also the only uses for now. I think it is worth the burning of that one record.

Whether you'd interpret the `0` byte as a version or not is something to define later I think? You can either define other versions with different formats, or map tlv straight onto the non-`0`-starting remainder of the data space.

Reply:
I don't see why you'd burn anything rather than just adopt the format already described in this PR and dedicate tag `0` to unstructured data (as done in [the inquisition PR](https://github.com/bitcoin-inquisition/bitcoin/pull/22)). In particular that answers two questions: how to have multiple data items in a single input (eg "ord", "image/jpeg", "[the jpeg]"), and how to use structured data / data with consensus meaning in future (use a different tag/have the first compressed int be multiple bytes).

To put `["ord", "text/plain;charset=utf8", "Hello, world"]` in the annex via the current proposal here, you'd encode:

```
   50  - annex
   03  - type 0, length 3
   6f7264 - "ord"
   17 - type 0, length 23
   746578742f706c61696e3b636861727365743d75746638 - text/plain;...
   0c - type 0, length 12
   48656c6c6f2c20776f726c64 - "Hello, world"
```

For data longer than 126 bytes, you'd prefix it with `7f` then the "CompressedInt" encoding of its length minus 127, eg `7f96e471` for 390,000 (encoded as 127 then 389873), and then you'd just include the 390kB of data directly. That's a total overhead of 13 witness bytes (3 bytes for the total annex witness size, 1 byte for the annex tag, 4 bytes for "ord" and its length, 5 bytes for the lengths of the mime type and data).

Reply:
I detailed another format that would optimize the unstructured annex data case more in https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-June/021756.html

In short: any annex is considered unstructured and does not incur any overhead at all, unless the annex starts with a magic byte which is to be repeated for unstructured data. Other values of the second byte are reserved for future extensions.

Reply:
> Other than that I think tlv is great. I proposed `0`-prefix for unstructured data to keep things as simple as possible initially without compromising future upgradeability, to avoid a potentially lengthy process to get to a standard for tlv.

If you don't like lengthy processes, don't extend them by proposing unnecessary alternatives when there's already a workable proposal?

Compared to this PR, what you propose adds overhead for encoding data that has consensus meaning, or where multiple items of data need to be included.

Reply:
>Compared to this PR, what you propose adds overhead for encoding data that has consensus meaning, or where multiple items of data need to be included.

It indeed depends on whether you want to space-optimize for structured or for unstructured data.

>If you don't like lengthy processes, don't extend them by proposing unnecessary alternatives when there's already a workable proposal?

I think it is important to explore alternatives. It maps out the design space and shows the trade-offs, which also exist in the case of tlv. Tlv isn't strictly better.

A "workable" proposal doesn't necessarily mean that it won't need to go through a lengthy process still. In my experience, starting with a simpler alternative can often expedite things and depending on future usage patterns it may even be the optimal choice.

Most important for me though is that the annex becomes usable in some form regardless of the exact space requirements. If you're saying that the tlv proposal can easily be guided through the process and enabled in policy for the next release, it's all good for me.

Reply:
> For large amounts of data, the overhead is indeed small. But there might be current or future use cases that only require smaller bits of unstructured data for which the overhead weighs more heavily. EDIT: For <127 bytes of unstructured data there is no overhead indeed in this proposal, so agree that this is hardly an argument.

If you have users leveraging smaller bits of unstructured data who cannot afford the TLV record bytes fee overhead cost, I think the economically rational position is to design a L2 system to uplift the unstructured data from on-chain to off-chain ?

Of course there is the question if you can maintain the accountability and visibility properties that your use-case is looking for with a L2 system relying on private state.

> It indeed depends on whether you want to space-optimize for structured or for unstructured data.

On the question of space-optimization, in my mind if the annex is used for economically space-sensitive use-case like channel factories or payment pools in the future, even few bytes of witness waste translates in raising the economic bar to afford access to those advanced Bitcoin payment systems.

Reply:
> I think it is important to explore alternatives.

It's valuable to explore alternatives if they potentially offer benefits; but this doesn't -- it just makes one use case slightly cheaper and other uses slightly more expensive. The cost to exploring alternatives is that it delays the entire process of making a decision, which was what you were complaining about in the first place.